### PR TITLE
DOCSP-37515 - Custom Types

### PR DESCRIPTION
**Pull request by mongoKart**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-37515>
Staging - https://preview-mongodbmongokart.gatsbyjs.io/pymongo/docsp-37515-custom-types/fundamentals/data-formats/custom-types/

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/fundamentals/data-formats/custom-types.txt [hunk 1]
```diff
@@ -1,348 +1,342 @@
...
-The type codec for our custom type simply needs to define how a
-``~decimal.Decimal`` instance can be converted into a
-``~bson.decimal128.Decimal128`` instance and vice-versa. Since we are
-interested in both encoding and decoding our custom type, we use the
-``TypeCodec`` base class to define our codec:
+After defining a custom type codec, you must add it to {+driver-short+}'s **type registry**,
+the list of types the driver can encode and decode.
+To do so, create an instance of the ``TypeRegistry`` class, passing in an instance
+of your type codec class inside a list. If you create multiple custom codecs, you can
+pass them all to the ``TypeRegistry`` constructor.
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:126]: S: Introduce code example for consistency w/ others
```suggestion
of your type codec class inside a list. If you create multiple custom codecs, you can
pass them all to the ``TypeRegistry`` constructor.

The following example adds the ``decimal_codec`` type codec to a type registry:

```

File: source/fundamentals/data-formats/custom-types.txt [hunk 2]
```diff
@@ -1,348 +1,342 @@
...
-Consider the situation where, in addition to encoding
-``~decimal.Decimal``, we also need to encode a type that subclasses
-``Decimal``. PyMongo does this automatically for types that inherit from
-Python types that are BSON-encodable by default, but the type codec system
-described above does not offer the same flexibility.
-
-Consider this subtype of ``Decimal`` that has a method to return its value as
-an integer:
+You might also need to encode one or more types that inherit from your custom type.
+Consider this subtype of the ``Decimal`` type, which contains a method to return its value
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:199]:
```suggestion
Consider the following subtype of the ``Decimal`` type, which contains a method to return its value
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 3]
```diff
@@ -1,348 +1,342 @@
...
-  information, it is impossible to discern which incoming
-  ``~bson.decimal128.Decimal128`` value needs to be decoded as ``Decimal``
-  and which needs to be decoded as ``DecimalInt``. This example only considers
-  the situation where a user wants to *encode* documents containing either
-  of these types.
-
-After creating a new codec options object and using it to get a collection
-object, we can seamlessly encode instances of ``DecimalInt``:
+To encode an instance of the ``DecimalInt`` class, you must define a type codec for
+the class. This type codec must inherit from the parent class's codec:
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:228]: S: clairfy the parent class's codec is DecimalCodec

```suggestion
the class. This type codec must inherit from the parent class's codec, ``DecimalCodec``, as shown in the following example:
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 4]
```diff
@@ -1,348 +1,342 @@
...
-    >>> newcoll = db.get_collection("new")
-    >>> doc = newcoll.find_one()
-    >>> type(doc["data"])
-    bson.binary.Binary
+      decimal_int_codec = DecimalIntCodec()
+      type_registry = TypeRegistry([decimal_codec, decimal_int_codec])
+      codec_options = CodecOptions(type_registry=type_registry)
 
+      collection = db.get_collection("test", codec_options=codec_options)
+      collection.drop()
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:255]: S: Remove collection.drop() since we aren't expecting them to run each of these examples iteratively. I don't think we want to make it seem like you have to drop the collection any time you do this

```suggestion
      collection = db.get_collection("test", codec_options=codec_options)

```

File: source/fundamentals/data-formats/custom-types.txt [hunk 5]
```diff
@@ -1,348 +1,342 @@
...
+   these values being decoded as ``Decimal``, not ``DecimalInt``.
 
-  For codebases requiring compatibility with both Python 2 and 3, type
-  decoders will have to be registered for both possible ``bson_type`` values.
+Define a Fallback Encoder
+-------------------------
 
+You can also register a **fallback encoder**, a callable to encode types not recognized
+by BSON and for which no type codec has been registered.
+Like the ``transform_python()`` method, the fallback encoder accepts an unencodable
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:276]: S: I don't think this is particularly relevant
```suggestion
The fallback encoder accepts an unencodable
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 6]
```diff
@@ -1,348 +1,342 @@
...
-Encoding Unknown Types
-^^^^^^^^^^^^^^^^^^^^^^
-
-In this example, we demonstrate how a fallback encoder can be used to save
-arbitrary objects to the database. We will use the the standard library's
-``pickle`` module to serialize the unknown types and so naturally, this
-approach only works for types that are picklable.
-
-We start by defining some arbitrary custom types:
+beforehand, you can them in cases where a ``TypeEncoder`` doesn't work.
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:337]:
```suggestion
beforehand, you can use them in cases where a ``TypeEncoder`` doesn't work.
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 7]
```diff
@@ -361,22 +355,21 @@ We start by defining some arbitrary custom types:
       def __repr__(self):
           return "MyNumberType(%s)" % (self.__value,)
 
-We also define a fallback encoder that pickles whatever objects it receives
-and returns them as ``~bson.binary.Binary`` instances with a custom
-subtype. The custom subtype, in turn, allows us to write a TypeDecoder that
+You can define a fallback encoder that pickles the objects it receives
+and returns them as ``Binary`` instances with a custom
+subtype. This custom subtype allows you to define a ``TypeDecoder`` class that
```
Comment (jordan-smith721) [inline source/fundamentals/data-formats/custom-types.txt:360]: This is likely work for another ticket somewhere down the line, but I think this example can probably be simplified. As someone unfamiliar with "pickling" a lot of this is hard to parse for me, and the code here takes a big jump in complexity compared to the rest of our pages. Might be worth investigating sometime after the standardization project

**Review by jordan-smith721 — CHANGES_REQUESTED** [review]
Nice job! I left a few suggestions and have some overall suggestions as well:

- I think we can simplify a lot of the code examples by removing `pprint` from them. I don't think it's accomplishing much and it adds extra lines to the code blocks
- I noticed this isn't in the TOC yet, is this going to be added to a section in another PR?

File: source/fundamentals/data-formats/custom-types.txt (hunk 4 above)
Comment (mongoKart) [reply source/fundamentals/data-formats/custom-types.txt:255]: yeah agreed

File: source/fundamentals/data-formats/custom-types.txt (hunk 5 above)
Comment (mongoKart) [reply source/fundamentals/data-formats/custom-types.txt:276]: i'll buy that

File: source/fundamentals/data-formats/custom-types.txt (hunk 7 above)
Comment (mongoKart) [reply source/fundamentals/data-formats/custom-types.txt:360]: agreed

**Review by jordan-smith721 — APPROVED** [review]
LGTM!

**Review by blink1073 — APPROVED** [review]
LGTM!
